utils/website/graphile-config/preset.md

```diff
@@ -1,49 +1,166 @@
+require() of ES modules is not supported.
+require() of /path/to/graphile.config.ts from /path/to/node_modules/graphile-config/dist/loadConfig.js is an ES module file as it is a .ts file whose nearest parent package.json contains "type": "module" which defines all .ts files in that package scope as ES modules.
+Instead change the requiring code to use import(), or remove "type": "module" from /path/to/package.json.
+```
+
+Or, in newer versions, an error saying unknown file extension:
+
+```js
+TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /path/to/graphile.config.ts
+```
+
+To solve this, use the experimental loaders API to add support for TS ESM via
+the `ts-node/esm` loader:
+
+```js
+export NODE_OPTIONS="$NODE_OPTIONS --loader ts-node/esm"
+```
+
+Then run your command again.
+
+## Preset scopes
+
+Libraries that use Graphile Config define scopes. Scopes are properties in a
+Graphile Config preset within which configuration options can be set. For
+example, [Postgraphile](https://postgraphile.org/) defines `postgraphile`,
+`pgServices` and other scopes. [Graphile Worker](https://worker.graphile.org/)
+defines a `worker` scope.
+
+Scopes are also used in plugins. Plugin authors 🔌 should see [Plugin](./plugin)
+for more details.
+
+## Preset composition
+
+Library authors create Graphile Config presets that allow consumers to easily
+use the library with default or recommended settings. The `extends` property in
+a preset allows library consumers to extend one or more presets with their own
+configuration values or plugins. Extending presets resolves values according to
+the [preset resolution algorithm](#preset-resolution).
+
+The following is a simple example of a Graphile Config preset that could be used
+to configure usage of [Graphile Worker](https://worker.graphile.org/):
+
+```ts title=graphile.config.ts
+import { WorkerPreset } from "graphile-worker";
+
+const preset: GraphileConfig.Preset = {
+  extends: [WorkerPreset],
+  worker: {
+    connectionString: "postgres:///my_db",
+  },
+};
+
+export default preset;
+```
+
+:::info
+
+As a library consumer, you can build and share presets with your collaborators
+or even extend 3rd party presets provided by the community. As with any 3rd
+party code, take caution to ensure you can trust the code you are importing.
+
+:::
+
+## Preset resolution
+
+Some libraries that use Graphile Config allow consumers to provide more than one
+preset. Presets may also compose (extend) other presets. When a library is
+passed a list of presets, it resolves the preset using the `ResolvePresets`
+algorithm below.
+
+tl;dr:
+
+- All the presets in `extends` are resolved in order (order is important!).
+- The plugins are merged as a set (each plugin will only be included once) and
+  sorted according to `before`/`after`.
+- The options are merged such that options specified last win.
+
+**ResolvePresets(presets):**
```

Comment:
Note to self: I'm not 100% sure this algorithm is correct now. Also: we should simplify it.